### PR TITLE
Update subctl diagnose to return exit code on failures

### LINKED ISSUE #1310: subctl diagnose to return relevant exit code on Submariner failures
`subctl diagnose` prints "Status.Error/Warnings" and exit normally, even if there's a connection failure.
This is a different response than already implemented by `subctl show` - that returns bad exit code on connection failures.

**What would you like to be added**:
Add relevant exit code for subctl diagnose.

**Why is this needed**:
common exit codes for subctl commands

### PR DESCRIPTION
**Pull request by maayanf24**
Fixes #1310

Signed-off-by: Maayan Friedman <maafried@redhat.com>

### CODE REVIEW

File: pkg/subctl/cmd/validate_cni.go [hunk 1]
```diff
@@ -87,7 +87,7 @@ func validateCNIInCluster(config *rest.Config, clusterName string, submariner *v
 			" Supported network plugins: %v\n", submariner.Status.NetworkPlugin, supportedNetworkPlugins)
 		status.QueueFailureMessage(message)
 		status.End(cli.Failure)
-		return
+		exitWithErrorMsg(message)
```
Comment (mangelajo) [inline pkg/subctl/cmd/validate_cni.go]: This will exit when running validate all, right?

Shouldn't we do this in L70. ?

And only after we have checked all the clusters, not only the first one.

**Review by mangelajo — CHANGES_REQUESTED** [review]
I'm not going exhaustively over all cases, but:

this changes the behavior, now as soon as there is any failure (even for diagnose all) it will stop and exit, we should save the errors and propagate upwards, while still continuing to run, and then, just before normal exit we can exit with error if any error happened otherwise, we cannot get a full diagnostic as soon as one single error appears.

File: pkg/subctl/cmd/validate_cni.go (hunk 1 above)
Comment (sridhargaddam) [reply pkg/subctl/cmd/validate_cni.go]: I agree with @mangelajo

Comment (tpantelis) [reply pkg/subctl/cmd/validate_cni.go]: +1.  We can have each validate function return a success/fail bool that is checked at the top-level.

**Comment by nyechiel** [comment]
(removing from the 0.9 board as this is already being tracked via the linked Issue)

**Review by nyechiel — APPROVED** [review]

File: pkg/subctl/cmd/validate_all.go [hunk 2]
```diff
@@ -37,8 +37,12 @@ func validateAll(cmd *cobra.Command, args []string) {
 	configs, err := getMultipleRestConfigs(kubeConfig, kubeContexts)
 	exitOnError("Error getting REST config for cluster", err)
 
+	validationStatus := true
+
 	for _, item := range configs {
-		validateK8sVersionInCluster(item.config, item.clusterName)
+		if !validateK8sVersionInCluster(item.config, item.clusterName) {
+			validationStatus = false
+		}
```
Comment (mangelajo) [inline pkg/subctl/cmd/validate_all.go]:
```suggestion
		validationStatus = validationStatus && validateK8sVersionInCluster(item.config, item.clusterName)
```

File: pkg/subctl/cmd/validate_all.go [hunk 3]
```diff
@@ -52,22 +56,40 @@ func validateAll(cmd *cobra.Command, args []string) {
 		status.End(cli.Success)
 		fmt.Println()
 
-		validateCNIInCluster(item.config, item.clusterName, submariner)
+		if !validateCNIInCluster(item.config, item.clusterName, submariner) {
+			validationStatus = false
+		}
```
Comment (mangelajo) [inline pkg/subctl/cmd/validate_all.go]: validationStatus = validationStatus && ...

File: pkg/subctl/cmd/validate_connections.go [hunk 4]
```diff
@@ -48,11 +50,16 @@ func validateConnections(cmd *cobra.Command, args []string) {
 			continue
 		}
 		status.End(cli.Success)
-		validateConnectionsInCluster(item.config, item.clusterName)
+		if !validateConnectionsInCluster(item.config, item.clusterName) {
+			validationStatus = false
```
Comment (mangelajo) [inline pkg/subctl/cmd/validate_connections.go]: validationStatus = validationStatus &&

**Review by mangelajo — CHANGES_REQUESTED** [review]

File: pkg/subctl/cmd/validate_all.go [hunk 5]
```diff
@@ -52,22 +56,40 @@ func validateAll(cmd *cobra.Command, args []string) {
...
 		fmt.Println()
 		fmt.Printf("Skipping tunnel firewall validation as it requires two kubeconfigs." +
 			" Please run \"subctl validate firewall tunnel\" command manually.\n")
 		fmt.Println()
 	}
+
+	if !validationStatus {
+		exitWithErrorMsg("Diagnose all return with an error")
```
Comment (tpantelis) [inline pkg/subctl/cmd/validate_all.go]: Any errors that occurred earlier were already outputted so I don't think we need an extra message here - we can just exit. Same elsewhere.

```suggestion
		os.Exit(1)
```

**Review by tpantelis — CHANGES_REQUESTED** [review]

File: pkg/subctl/cmd/validate_cni.go [hunk 6]
```diff
@@ -207,7 +218,7 @@ func validateCalicoIPPoolsIfCalicoCNI(config *rest.Config) {
 		}
 	}
 
-	status.End(status.ResultFromMessages())
```
Comment (mangelajo) [inline pkg/subctl/cmd/validate_cni.go:210]: Should we really remove this status.End? or keep it (in addition to the return true?)

Comment (tpantelis) [reply pkg/subctl/cmd/validate_cni.go:210]: We need to keep this and also return false if there were any errors. 

```
result := status.ResultFromMessages()
status.End(result)

return result != cli.Failure
```

File: pkg/subctl/cmd/validate_deployment.go [hunk 7]
```diff
@@ -224,7 +242,7 @@ func checkPodsStatus(k8sClient kubernetes.Interface, operatorNamespace string) b
 			message := fmt.Sprintf("Pod %q is not running. (current state is %v)", pod.Name, pod.Status.Phase)
 			status.QueueFailureMessage(message)
 			status.End(cli.Failure)
-			return false
+			exitWithErrorMsg(message)
```
Comment (tpantelis) [inline pkg/subctl/cmd/validate_deployment.go]: Why was this changed to exit here?

File: pkg/subctl/cmd/validate_fw_vxlan.go [hunk 8]
```diff
@@ -59,47 +61,56 @@ func validateFirewallVxLANConfig(cmd *cobra.Command, args []string) {
...
-		return
+		return false
 	}
 
 	gateways := getGatewaysResource(config)
 	if gateways == nil {
-		status.QueueWarningMessage("There are no gateways detected on the cluster.")
-		return
+		message := "There are no gateways detected on the cluster."
+		status.QueueWarningMessage(message)
```
Comment (tpantelis) [inline pkg/subctl/cmd/validate_fw_vxlan.go]: nit but why was this changed to store message in a var?

```suggestion
		status.QueueWarningMessage("There are no gateways detected on the cluster.")
```

File: pkg/subctl/cmd/validate_fw_vxlan.go [hunk 9]
```diff
@@ -59,47 +61,56 @@ func validateFirewallVxLANConfig(cmd *cobra.Command, args []string) {
...
+		return false
 	}
 
-	if len(gateways.Items[0].Status.Connections) == 0 {
-		status.QueueWarningMessage("There are no active connections to remote clusters.")
-		return
+	if len(gateways.Items) == 0 || len(gateways.Items[0].Status.Connections) == 0 {
+		message := "There are no active connections to remote clusters."
+		status.QueueWarningMessage(message)
```
Comment (tpantelis) [inline pkg/subctl/cmd/validate_fw_vxlan.go]: same here

**Review by tpantelis — CHANGES_REQUESTED** [review]
I think we should also run `subctl diagnose`  as part of the E2E test as a sanity check.

File: pkg/subctl/cmd/validate_fw_vxlan.go [hunk 10]
```diff
@@ -59,47 +62,52 @@ func validateFirewallVxLANConfig(cmd *cobra.Command, args []string) {
...
 	clientSet, err := kubernetes.NewForConfig(config)
 	if err != nil {
 		message := fmt.Sprintf("Error creating API server client: %s", err)
 		status.QueueFailureMessage(message)
-		return
+		return false
 	}
 
 	gateways := getGatewaysResource(config)
 	if gateways == nil {
```
Comment (tpantelis) [inline pkg/subctl/cmd/validate_fw_vxlan.go]: `if gateways == nil || len(gateways.Items) == 0 { `

File: pkg/subctl/cmd/validate_fw_vxlan.go [hunk 11]
```diff
@@ -59,47 +62,52 @@ func validateFirewallVxLANConfig(cmd *cobra.Command, args []string) {
...
 		status.QueueWarningMessage("There are no gateways detected on the cluster.")
-		return
+		return false
 	}
 
-	if len(gateways.Items[0].Status.Connections) == 0 {
+	if len(gateways.Items) == 0 || len(gateways.Items[0].Status.Connections) == 0 {
```
Comment (tpantelis) [inline pkg/subctl/cmd/validate_fw_vxlan.go]:
```suggestion
	if len(gateways.Items[0].Status.Connections) == 0 {
```

**Review by tpantelis — CHANGES_REQUESTED** [review]

File: pkg/subctl/cmd/validate_fw_vxlan.go (hunk 11 above)
Comment (dfarrell07) [inline pkg/subctl/cmd/validate_fw_vxlan.go]: Was adding this new `len(gateways.Items) == 0` check intentional for this change? If so, it might be good to document it in the commit message.

Comment (dfarrell07) [reply pkg/subctl/cmd/validate_fw_vxlan.go]: Ah, ditto to the review I was just working on.

Comment (maayanf24) [reply pkg/subctl/cmd/validate_fw_vxlan.go]: it will prevent a panic when there is a gateway object with no items

**Review by mangelajo — APPROVED** [review]

File: pkg/subctl/cmd/validate_fw_vxlan.go (hunk 11 above)
Comment (mangelajo) [reply pkg/subctl/cmd/validate_fw_vxlan.go]: let's keep it simple and add this now.

**Review by sridhargaddam — APPROVED** [review]
thanks @maayanf24

**Review by tpantelis — APPROVED** [review]
